### PR TITLE
Introduce benchmarking using Criterion.

### PR DESCRIPTION
**Pull request by Hoverbear**
Starts on #109.

This demonstrates Criterion, ways to bench cleanly, and documentation how to use it.

### CODE REVIEW

File: benches/benches.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#![allow(dead_code)] // Due to criterion we need this to avoid warnings.
+
+#[macro_use]
+extern crate criterion;
+extern crate raft;
+
+mod suites;
+use suites::*;
+
+criterion_main!(
```
Comment (siddontang) [inline benches/benches.rs]: the default sample size for the criterion is 100, maybe we can use a smaller sample size like 10 to reduce the benchmark time.

**Comment by siddontang** [comment]
CI failed @Hoverbear

**Comment by siddontang** [comment]
PTAL @breeswish

**Comment by siddontang** [comment]
LGTM

File: benches/benches.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+use criterion::Criterion;
+use std::time::Duration;
+
+mod suites;
+
+pub const DEFAULT_RAFT_SETS: [(usize, usize); 4] = [(0, 0), (3, 1), (5, 2), (7, 3)];
+
+fn main() {
+    criterion::init_logging();
+    let _ = env_logger::try_init();
```
Comment (breeswish) [inline benches/benches.rs]: Seems to be duplicated? criterion::init_logging will use a SimpleLogger to initialize `log`.

File: src/eraftpb.rs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 // https://github.com/Manishearth/rust-clippy/issues/702
 #![allow(unknown_lints)]
-#![allow(clippy)]
+#![cfg_attr(feature = "cargo-clippy", allow(clippy::all))]
```
Comment (breeswish) [inline src/eraftpb.rs]: This is a generated file, maybe we should not edit it?

Comment (Hoverbear) [reply src/eraftpb.rs]: That would mean blocking this and all future raft-rs PRs :( It's part of #112

File: benches/benches.rs (hunk 2 above)
Comment (Hoverbear) [reply benches/benches.rs]: Oh nice catch. :)

**Comment by Hoverbear** [comment]
@breeswish Thanks, I resolved your comment. PTAL again?

**Pull request merged**